src/tickit/utils/configuration/configurable.py

```diff
@@ -1,67 +1,97 @@
-is_tagged_union: Dict[Type[Any], bool] = DefaultDict(lambda: False)
-
-
-def as_tagged_union(cls: Cls) -> Cls:
-    """A decorator to make a config base class that can deserialize aliased sub-classes.
-
-    A decorator which makes a config class the root of a tagged union of sub-classes
-    allowing for serialization and deserialization of config trees by class alias. The
-    function registers both an apischema serialization and an apischema deserialization
-    conversion for the base class which perform lookup based on a tagged union of
-    aliased sub-classes.
-
-    Args:
-        cls (Cls): The config base class.
-
-    Returns:
-        Cls: The modified config base class.
-    """
-
-    def deserialization() -> (
-        Conversion
-    ):  # This will only be used if we want to generate a json schema (which we will)
-        annotations: Dict[str, Any] = {}
-        deserialization_namespace: Dict[str, Any] = {"__annotations__": annotations}
-        for sub in rec_subclasses(cls):
-            fullname = sub.__module__ + "." + sub.__name__
-            annotations[fullname] = Tagged[sub]  # type: ignore
-        deserialization_union = type(
-            cls.__name__ + "TaggedUnion",
-            (TaggedUnion,),
-            deserialization_namespace,
-        )
-        return Conversion(
-            lambda obj: get_tagged(obj)[1], source=deserialization_union, target=cls
-        )
-
-    deserializer(lazy=deserialization, target=cls)
-    is_tagged_union[cls] = True
-    return cls
+from dataclasses import field
+from importlib import import_module
+from typing import Any, Callable, Literal, Optional, Type, Union
+
+from pydantic.v1 import BaseConfig, Field, ValidationError, create_model, parse_obj_as
+from pydantic.v1.error_wrappers import ErrorWrapper
+
+
+def as_tagged_union(
+    super_cls: Optional[Union[Type, Callable[[Type], Type]]] = None,
+    *,
+    discriminator: str = "type",
+    config: Optional[Type[BaseConfig]] = None,
+) -> Union[Type, Callable[[Type], Type]]:
+    def wrap(cls):
+        return _as_tagged_union(cls, discriminator, config)
+
+    # Work out if the call was @discriminated_union_of_subclasses or
+    # @discriminated_union_of_subclasses(...)
+    if super_cls is None:
+        return wrap
+    else:
+        return wrap(super_cls)
+
+
+def qualified_class_name(klass):
+    module = klass.__module__
+    if module == "builtins":
+        return klass.__qualname__  # avoid outputs like 'builtins.str'
+    return module + "." + klass.__qualname__
+
+
+def _as_tagged_union(
+    super_cls: Type,
+    discriminator: str,
+    config: Optional[Type[BaseConfig]] = None,
+) -> Union[Type, Callable[[Type], Type]]:
+    super_cls._ref_classes = set()
+    super_cls._model = None
+    setattr(super_cls, discriminator, field())
+
+    def _load_module_with_type(values: dict[str, str]) -> None:
+        fullname = values.get(discriminator)
+        assert fullname
```

Comment:
Would be better as if/raise to prevent relying on assert. And to give better errors